File: Android/TagMaster/src/main/java/depollsoft/tagmaster/CustomListsModel.kt
package depollsoft.tagmaster

import com.google.firebase.auth.ktx.auth
import com.google.firebase.firestore.FieldValue
import com.google.firebase.firestore.SetOptions
import com.google.firebase.firestore.ktx.firestore
import com.google.firebase.ktx.Firebase
import depollsoft.lib.util.Preferences
import java.util.UUID

/**
 * Manages metadata for custom tag lists (name, order).
 * Built-in lists (favorite, teachable) have fixed metadata.
 * Custom list metadata is stored in SharedPreferences and synced to
 * Firestore at users/{uid}/listMeta.
 */
object CustomListsModel {
    private const val LIST_META_KEY = "tagmaster.listMeta"

    private var metadata: MutableMap<String, ListMetadata> = run {
        val stored: Map<String, Map<String, Any>>? = Preferences.get(LIST_META_KEY)
        val result = mutableMapOf<String, ListMetadata>()
        stored?.forEach { (key, map) ->
            val name = map["name"] as? String ?: key
            val order = (map["order"] as? Number)?.toInt() ?: 0
            result[key] = ListMetadata(key, name, order, isBuiltIn = false)
        }
        result
    }

    private var shouldStore = true

    /**
     * For testing: disable Firestore storage.
     */
    @JvmStatic
    fun setTestMode(enabled: Boolean) {
        shouldStore = !enabled
    }

    /**
     * Returns all lists: built-in first, then custom lists sorted by order.
     */
    fun allLists(): List<ListMetadata> {
        val builtIn = listOf(ListMetadata.favorite, ListMetadata.teachable)
        val custom = metadata.values.sortedBy { it.order }
        return builtIn + custom
    }

    /**
     * Returns only custom lists sorted by order.
     */
    fun customLists(): List<ListMetadata> {
        return metadata.values.sortedBy { it.order }
    }

    /**
     * Gets metadata for a list by key, or null if not found.
     * Returns built-in metadata for favorite/teachable keys.
     */
    fun getMetadata(key: String): ListMetadata? {
        return when (key) {
            ListMetadata.FAVORITE_KEY -> ListMetadata.favorite
            ListMetadata.TEACHABLE_KEY -> ListMetadata.teachable
            else -> metadata[key]
        }
    }

    /**
     * Creates a new custom list with the given name.
     * Returns the generated key.
     */
    fun createList(name: String): String {
        val key = UUID.randomUUID().toString()
        val order = (metadata.values.maxOfOrNull { it.order } ?: 1) + 1
        val meta = ListMetadata(key, name, order, isBuiltIn = false)
        metadata[key] = meta
        // Create the backing ListModel so it's ready to use
        ListModel(key)
        storeLocally()
        storeToFirestore(key, meta)
        return key
    }

    /**
     * Renames a custom list. No-op for built-in lists.
     */
    fun renameList(key: String, newName: String) {
        val existing = metadata[key] ?: return
        metadata[key] = existing.copy(name = newName)
        storeLocally()
        storeToFirestore(key, metadata[key]!!)
    }

    /**
     * Deletes a custom list and its tag IDs. No-op for built-in lists.
     */
    fun deleteList(key: String) {
        if (key == ListMetadata.FAVORITE_KEY || key == ListMetadata.TEACHABLE_KEY) return
        metadata.remove(key)
        ListModel.deleteList(key)
        storeLocally()
        deleteFromFirestore(key)
    }

    private fun storeLocally() {
        val toStore = mutableMapOf<String, Map<String, Any>>()
        metadata.forEach { (key, meta) ->
            toStore[key] = mapOf("name" to meta.name, "order" to meta.order)
        }
        Preferences.set(LIST_META_KEY, toStore)
    }

    private fun storeToFirestore(key: String, meta: ListMetadata) {
        if (!shouldStore) return
        val user = Firebase.auth.currentUser ?: return
        val userDoc = Firebase.firestore.document("users/${user.uid}")
        userDoc.set(
            mapOf("listMeta" to mapOf(key to mapOf("name" to meta.name, "order" to meta.order))),
            SetOptions.mergeFields("listMeta.${key}")
        )
    }

    private fun deleteFromFirestore(key: String) {
        if (!shouldStore) return
        val user = Firebase.auth.currentUser ?: return
        val userDoc = Firebase.firestore.document("users/${user.uid}")
        userDoc.set(
            mapOf(
                "lists" to mapOf(key to FieldValue.delete()),
                "listMeta" to mapOf(key to FieldValue.delete())
            ),
            SetOptions.mergeFields("lists.${key}", "listMeta.${key}")
        )
    }

    /**
     * Syncs metadata from a Firestore snapshot data map.
     * Called from ListModel.fromFirestore().
     */
    fun fromFirestore(data: Map<*, *>?) {
        if (data == null) return
        shouldStore = false
        try {
            val newMeta = mutableMapOf<String, ListMetadata>()
            data.forEach { (key, value) ->
                if (key !is String) return@forEach
                val map = value as? Map<*, *> ?: return@forEach
                val name = map["name"] as? String ?: key
                val order = (map["order"] as? Number)?.toInt() ?: 0
                newMeta[key] = ListMetadata(key, name, order, isBuiltIn = false)
            }
            metadata = newMeta
            storeLocally()
        } finally {
            shouldStore = true
        }
    }

    /**
     * Returns metadata as a map suitable for Firestore storage.
     */
    fun toFirestoreMap(): Map<String, Map<String, Any>> {
        val result = mutableMapOf<String, Map<String, Any>>()
        metadata.forEach { (key, meta) ->
            result[key] = mapOf("name" to meta.name, "order" to meta.order)
        }
        return result
    }

    /**
     * Resets all state. For testing only.
     */
    @JvmStatic
    fun resetForTesting() {
        metadata = mutableMapOf()
    }
}

File: iOS/tagmaster/tagmaster/CustomListsModel.swift
//
//  CustomListsModel.swift
//  tagmaster
//
//  Created by David Poll on 2025.
//  Copyright © 2025 DepollSoft. All rights reserved.
//

import Foundation
import Firebase

/// Metadata for a tag list. Built-in lists (favorite, teachable) have fixed
/// display names and cannot be renamed or deleted.
struct ListMetadata {
    static let favoriteKey = "favorite"
    static let teachableKey = "teachable"

    static let favorite = ListMetadata(key: favoriteKey, name: "Favorites", order: 0, isBuiltIn: true)
    static let teachable = ListMetadata(key: teachableKey, name: "Teachable Tags", order: 1, isBuiltIn: true)

    let key: String
    let name: String
    let order: Int
    let isBuiltIn: Bool
}

/// Manages metadata for custom tag lists (name, order).
/// Built-in lists (favorite, teachable) have fixed metadata.
/// Custom list metadata is stored in UserDefaults and synced to Firestore.
/// Mirrors the Android `CustomListsModel` pattern.
class CustomListsModel {

    private static let listMetaKey = "tagmaster.listMeta"

    private static var metadata: [String: ListMetadata] = {
        guard let stored = UserDefaults.standard.dictionary(forKey: listMetaKey) as? [String: [String: Any]] else {
            return [:]
        }
        var result: [String: ListMetadata] = [:]
        for (key, map) in stored {
            let name = map["name"] as? String ?? key
            let order = (map["order"] as? NSNumber)?.intValue ?? 0
            result[key] = ListMetadata(key: key, name: name, order: order, isBuiltIn: false)
        }
        return result
    }()

    private static var shouldStore = true

    // MARK: - Test Mode

    static func setTestMode(_ enabled: Bool) {
        shouldStore = !enabled
    }

    static func resetForTesting() {
        metadata = [:]
    }

    // MARK: - Queries

    /// Returns all lists: built-in first, then custom lists sorted by order.
    static func allLists() -> [ListMetadata] {
        let builtIn = [ListMetadata.favorite, ListMetadata.teachable]
        let custom = metadata.values.sorted { $0.order < $1.order }
        return builtIn + custom
    }

    /// Returns only custom lists sorted by order.
    static func customLists() -> [ListMetadata] {
        return metadata.values.sorted { $0.order < $1.order }
    }

    /// Gets metadata for a list by key, or nil if not found.
    /// Returns built-in metadata for favorite/teachable keys.
    static func getMetadata(_ key: String) -> ListMetadata? {
        switch key {
        case ListMetadata.favoriteKey:
            return ListMetadata.favorite
        case ListMetadata.teachableKey:
            return ListMetadata.teachable
        default:
            return metadata[key]
        }
    }

    // MARK: - Mutators

    /// Creates a new custom list with the given name. Returns the generated key.
    @discardableResult
    static func createList(_ name: String) -> String {
        let key = UUID().uuidString
        let order = (metadata.values.map { $0.order }.max() ?? 1) + 1
        let meta = ListMetadata(key: key, name: name, order: order, isBuiltIn: false)
        metadata[key] = meta
        // Create the backing ListModel so it's ready to use
        _ = ListModel.get(key)
        storeLocally()
        storeToFirestore(key: key, meta: meta)
        return key
    }

    /// Renames a custom list. No-op for built-in lists.
    static func renameList(_ key: String, newName: String) {
        guard let existing = metadata[key] else { return }
        let updated = ListMetadata(key: key, name: newName, order: existing.order, isBuiltIn: false)
        metadata[key] = updated
        storeLocally()
        storeToFirestore(key: key, meta: updated)
    }

    /// Deletes a custom list and its tag IDs. No-op for built-in lists.
    static func deleteList(_ key: String) {
        guard key != ListMetadata.favoriteKey && key != ListMetadata.teachableKey else { return }
        metadata.removeValue(forKey: key)
        ListModel.deleteList(key)
        storeLocally()
        deleteFromFirestore(key: key)
    }

    // MARK: - Local Persistence

    private static func storeLocally() {
        var toStore: [String: [String: Any]] = [:]
        for (key, meta) in metadata {
            toStore[key] = ["name": meta.name, "order": meta.order]
        }
        UserDefaults.standard.set(toStore, forKey: listMetaKey)
    }

    // MARK: - Firestore

    private static func storeToFirestore(key: String, meta: ListMetadata) {
        guard shouldStore else { return }
        guard let user = Auth.auth().currentUser else { return }
        let userDoc = Firestore.firestore().document("users/\(user.uid)")
        userDoc.setData(
            ["listMeta": [key: ["name": meta.name, "order": meta.order]]],
            mergeFields: ["listMeta.\(key)"]
        )
    }

    private static func deleteFromFirestore(key: String) {
        guard shouldStore else { return }
        guard let user = Auth.auth().currentUser else { return }
        let userDoc = Firestore.firestore().document("users/\(user.uid)")
        userDoc.setData(
            [
                "lists": [key: FieldValue.delete()],
                "listMeta": [key: FieldValue.delete()]
            ],
            mergeFields: ["lists.\(key)", "listMeta.\(key)"]
        )
    }

    /// Syncs metadata from a Firestore snapshot. Gracefully handles nil (missing listMeta).
    static func fromFirestore(_ data: [String: Any]?) {
        guard let data = data else { return }
        shouldStore = false
        defer {
            shouldStore = true
        }
        var newMeta: [String: ListMetadata] = [:]
        for (key, value) in data {
            guard let map = value as? [String: Any] else { continue }
            let name = map["name"] as? String ?? key
            let order = (map["order"] as? NSNumber)?.intValue ?? 0
            newMeta[key] = ListMetadata(key: key, name: name, order: order, isBuiltIn: false)
        }
        metadata = newMeta
        storeLocally()
    }

    /// Returns metadata as a map suitable for Firestore storage.
    static func toFirestoreMap() -> [String: [String: Any]] {
        var result: [String: [String: Any]] = [:]
        for (key, meta) in metadata {
            result[key] = ["name": meta.name, "order": meta.order]
        }
        return result
    }
}
